zkvm/src/types.rs

```diff
@@ -156,6 +159,35 @@ impl ScalarWitness {
     }
 }
 
+impl Neg for ScalarWitness {
+    type Output = ScalarWitness;
+
+    fn neg(self) -> ScalarWitness {
+        match self {
+            ScalarWitness::Integer(a) => ScalarWitness::Integer(-a),
+            ScalarWitness::Scalar(a) => ScalarWitness::Scalar(-a),
+        }
+    }
+}
+impl Add for ScalarWitness {
+    type Output = ScalarWitness;
+
+    fn add(self, rhs: ScalarWitness) -> ScalarWitness {
+        match (self, rhs) {
+            (ScalarWitness::Integer(a), ScalarWitness::Integer(b)) => match a + b {
+                Some(res) => ScalarWitness::Integer(res),
+                None => ScalarWitness::Scalar(
+                    <SignedInteger as Into<Scalar>>::into(a)
```

Comment:
Seems like it'd be worth it defining `to_scalar(self) -> Scalar` and using it inside Into<Scalar> impl and here to make things much clearer.